jujugui/static/gui/src/app/app.js

```diff
@@ -1187,78 +1196,153 @@ Browser: ${navigator.userAgent}`
     @param {Boolean} showIndicator
   */
-  _renderDragOverNotification(showIndicator = true) {
-    this.topology.fadeHelpIndicator(showIndicator);
-    ReactDOM.render(
-      <ExpandingProgress />,
-      document.getElementById('drag-over-notification-container'));
+  _generateDragOverNotification(showIndicator = true) {
+    this.props.topology.fadeHelpIndicator(showIndicator);
+    return (
+      <ExpandingProgress />);
   }
+
   /**
     Hide the drag notifications.
   */
   _hideDragOverNotification() {
-    this.topology.fadeHelpIndicator(false);
-    ReactDOM.unmountComponentAtNode(
-      document.getElementById('drag-over-notification-container'));
+    this.props.topology.fadeHelpIndicator(false);
+    // TODO: use state to handle hiding this
+    // ReactDOM.unmountComponentAtNode(
+    //   document.getElementById('drag-over-notification-container'));
   }
+
   /**
     Renders the zoom component to the page in the designated element.
   */
-  _renderZoom() {
-    ReactDOM.render(
-      <Zoom
-        topo={this.topology.topo} />,
-      document.getElementById('zoom-container'));
+  _generateZoom() {
+    return (
+      <div id="zoom-container">
+        <Zoom
+          topo={this.props.topology.topo} />
+      </div>);
   }
+
   /**
-    Render the react components.
+    Generate the cookie notice.
   */
-  _renderComponents() {
-    // Update the react views on database change
-    this._renderEnvSizeDisplay(
-      this.db.services.size(),
-      this.db.machines.filterByParent().length
-    );
-    this._renderDeploymentBar();
-    this._renderModelActions();
-    this._renderProviderLogo();
-    this._renderZoom();
-    this._renderBreadcrumb();
-    this._renderHeaderSearch();
-    this._renderHeaderHelp();
-    this._renderHeaderLogo();
-    const gui = this.state.current.gui;
-    if (!gui || (gui && !gui.inspector)) {
-      this._renderAddedServices();
+  _generateCookieNotice() {
+    if (this.props.applicationConfig.GTM_enabled) {
+      return cookieUtil.check(document, this.props.appState);
     }
   }
+
+  render() {
+    return (
+      <div>
+        <div className="header-banner header-banner--left">
+          {this._generateHeaderLogo()}
+          {this._generateBreadcrumb()}
+          {this._generateModelActions()}
+          {this._generateProviderLogo()}
+        </div>
+        <div className="header-banner header-banner--right">
+          <ul className="header-banner__list--right">
+            <li className="header-banner__list-item"
+              id="maas-server"
+              style={{ display: 'none' }} >
+              <a className="header-banner__link"
+                href=""
+                target="_blank">
+                MAAS UI
+              </a>
+            </li>
+            {this._generateHeaderSearch()}
+            {this._generateHeaderHelp()}
+            {this._generateUserMenu()}
+          </ul>
+        </div>
+        {this._generateZoom()}
```

Comment:
The index container nodes (e.g. `<div id="charmbrowser-container"></div>`) are removed from the two templaces and are replaced with these component generators, just like we handle conditional components elsewhere.